File: src/algorithms/interview/arrays/ArrayManipulation.java
package algorithms.interview.arrays;

import java.io.IOException;
import java.util.Scanner;

/**
 * Created by Mohammed Fathy @ 8/7/2018
 * dev.mfathy@gmail.com
 * <p>
 */
// TODO: 8/7/2018 Not fully solved
public class ArrayManipulation {

    // Complete the arrayManipulation function below.
    static long arrayManipulation(int n, int[][] queries) {

        long[] zArray = new long[n];

        long maxValue = 0;
        for (int[] query : queries) {
            int aA = query[0];
            int bB = query[1];
            int kK = query[2];


            for (int j = 0; j < zArray.length; j++) {
                if (j + 1 >= aA && j + 1 <= bB) {
                    zArray[j] = zArray[j] + kK;
                    if (maxValue < zArray[j]) maxValue = zArray[j];
                }
            }

        }

        return maxValue;

    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
//        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] nm = scanner.nextLine().split(" ");

        int n = Integer.parseInt(nm[0]);

        int m = Integer.parseInt(nm[1]);

        int[][] queries = new int[m][3];

        for (int i = 0; i < m; i++) {
            String[] queriesRowItems = scanner.nextLine().split(" ");
            scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

            for (int j = 0; j < 3; j++) {
                int queriesItem = Integer.parseInt(queriesRowItems[j]);
                queries[i][j] = queriesItem;
            }
        }

        long result = arrayManipulation(n, queries);

        System.out.println(String.valueOf(result));

        scanner.close();
    }
}

File: src/algorithms/interview/arrays/ArrayManipulation.kt
/**
 * Created by Mohammed Fathy @ 01/02/2019
 * dev.mfathy@gmail.com
 * https://www.hackerrank.com/challenges/crush?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=arrays
 * Explanation https://codereview.stackexchange.com/questions/185320/hackerrank-array-manipulation
 *
 */

import java.util.*

// Complete the arrayManipulation function below.
fun arrayManipulation(n: Int, queries: Array<Array<Int>>): Long {
    val sumAr = LongArray(n+1)
    var maxSum = 0L
    if (n in 3..100000000){
        queries.forEach { query ->
            val a = query[0] - 1
            val b = query[1] - 1
            val k = query[2].toLong()
            sumAr[a] += k
            if (b + 1 < n) {
                sumAr[b+1] -= k
            }
        }

        maxSum = sumAr[0]
        var internSum = 0L
        for (i in 0 .. n) {
            internSum += sumAr[i]
            if (internSum > maxSum){
                maxSum = internSum
            }
        }
    }

    return maxSum
}

fun main(args: Array<String>) {
    val scan = Scanner(System.`in`)

    val nm = scan.nextLine().split(" ")

    val n = nm[0].trim().toInt()

    val m = nm[1].trim().toInt()

    val queries = Array<Array<Int>>(m, { Array<Int>(3, { 0 }) })

    for (i in 0 until m) {
        queries[i] = scan.nextLine().split(" ").map{ it.trim().toInt() }.toTypedArray()
    }

    val result = arrayManipulation(n, queries)

    println(result)
}